File: Sources/ClaudeRelayKit/Protocol/PairingURL.swift
import Foundation

/// The `coderelay://pair?host=&port=&tls=&code=` deep link produced by
/// `claude-relay setup` and consumed by the apps.
///
/// Parsing and validation live here, in the shared kit, so the server-side
/// producer and all three clients agree on exactly what a valid pairing link
/// is — and so hostile input is rejected in one tested place.
public struct PairingURL: Equatable, Sendable {
    /// The URL scheme, renamed from `clauderelay` with the ClaudeRelay →
    /// CodeRelay rebrand.
    ///
    /// Deliberately a hard cutover, not a dual-scheme deprecation: a pairing
    /// code lives five minutes and a session link is generated on demand, so
    /// there is no persisted corpus of old URLs to honour. The cost is that
    /// server and clients must be updated together — a QR from an older
    /// `claude-relay setup` will not open in a client built after this change.
    public static let scheme = "coderelay"
    public static let host = "pair"

    public let host: String
    public let port: UInt16
    public let useTLS: Bool
    /// Always normalized (uppercase, no hyphen).
    public let code: String

    public init(host: String, port: UInt16, useTLS: Bool, code: String) {
        self.host = host
        self.port = port
        self.useTLS = useTLS
        self.code = code
    }

    public var urlString: String {
        var components = URLComponents()
        components.scheme = Self.scheme
        components.host = Self.host
        components.queryItems = [
            URLQueryItem(name: "host", value: host),
            URLQueryItem(name: "port", value: String(port)),
            URLQueryItem(name: "tls", value: useTLS ? "1" : "0"),
            URLQueryItem(name: "code", value: code)
        ]
        return components.url?.absoluteString ?? ""
    }

    /// The WebSocket URL a client should dial to redeem this code.
    public var wsURL: URL? {
        URL(string: "\(useTLS ? "wss" : "ws")://\(host):\(port)")
    }

    public init?(string: String) {
        guard let url = URL(string: string) else { return nil }
        self.init(url: url)
    }

    /// Whether `host` can form a usable `ws://host:port` URL.
    ///
    /// Hoisted out of `init?(url:)` so a producer can validate an
    /// operator-supplied host *before* it has a port, a TLS flag, or a minted
    /// code to build a full URL with. Both the producer and the parser ask this
    /// one function, so "what is a valid pairing host" has a single answer.
    public static func isValidHost(_ host: String) -> Bool {
        let trimmed = host.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty else { return false }
        // Reject RFC 3986-illegal host characters before handing the string to
        // URL(string:), which is lenient about some of them.
        guard trimmed.rangeOfCharacter(from: CharacterSet(charactersIn: " /?#@")) == nil else {
            return false
        }
        // The port here is a stand-in: host validity does not depend on it, and
        // any in-range port exercises the same URL parse.
        return URL(string: "ws://\(trimmed):1") != nil
    }

    public init?(url: URL) {
        guard url.scheme?.lowercased() == Self.scheme,
              url.host?.lowercased() == Self.host,
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              let items = components.queryItems
        else { return nil }

        func value(_ name: String) -> String? {
            items.first { $0.name == name }?.value
        }

        guard let rawHost = value("host")?.trimmingCharacters(in: .whitespaces),
              !rawHost.isEmpty,
              let rawPort = value("port"), let port = UInt16(rawPort), port >= 1,
              let rawCode = value("code"), let code = PairingCode.normalize(rawCode)
        else { return nil }

        let useTLS = value("tls") == "1"

        // Reject anything that cannot form a usable ws:// URL (spaces, slashes,
        // other RFC 3986-illegal host characters).
        guard Self.isValidHost(rawHost) else { return nil }

        self.host = rawHost
        self.port = port
        self.useTLS = useTLS
        self.code = code
    }
}

File: ClaudeRelayAndroid/core-protocol/src/main/kotlin/relay/protocol/PairingURL.kt
package relay.protocol

import java.net.URI
import java.net.URLDecoder

/**
 * Kotlin mirror of Swift PairingURL. Pure JVM — no android.net.Uri.
 *
 * The `clauderelay://pair?host=&port=&tls=&code=` deep link produced by
 * `claude-relay setup` and consumed by the apps.
 *
 * Parsing and validation live here, in the shared protocol module, so the
 * server-side producer and all three clients agree on exactly what a valid
 * pairing link is — and so hostile input is rejected in one tested place.
 */
data class PairingURL(
    val host: String,
    val port: Int,
    val useTLS: Boolean,
    /** Always normalized (uppercase, no hyphen). */
    val code: String,
) {
    /** The WebSocket URL a client should dial to redeem this code. */
    val wsUrl: String get() = "${if (useTLS) "wss" else "ws"}://$host:$port"

    companion object {
        const val SCHEME = "clauderelay"
        const val HOST = "pair"

        /**
         * Parses a pairing URL string, or returns null if it's invalid.
         *
         * Validates:
         * - scheme must be "clauderelay"
         * - host/authority must be "pair"
         * - query must have host (non-empty, trimmed), port (1..65535), code (via PairingCode.normalize)
         * - tls="1" → useTLS=true
         * - rejects RFC-3986-illegal host chars (space / ? # @)
         */
        fun parse(input: String): PairingURL? {
            val uri = try { URI(input) } catch (_: Exception) { return null }

            // Validate scheme
            if (uri.scheme?.lowercase() != SCHEME) return null

            // URI puts "pair" in authority/host for clauderelay://pair?...
            val authority = (uri.host ?: uri.authority)?.lowercase()
            if (authority != HOST) return null

            // Parse query parameters
            val query = uri.rawQuery ?: return null
            val params = parseQueryParams(query) ?: return null

            // Extract and validate host
            val host = params["host"]?.trim()
            if (host.isNullOrEmpty()) return null

            // Validate port
            val port = params["port"]?.toIntOrNull() ?: return null
            if (port < 1 || port > 65535) return null

            // Validate and normalize code
            val code = params["code"]?.let { PairingCode.normalize(it) } ?: return null

            // Parse TLS flag
            val useTLS = params["tls"] == "1"

            // Reject RFC-3986-illegal host characters
            if (host.any { it == ' ' || it == '/' || it == '?' || it == '#' || it == '@' }) {
                return null
            }

            return PairingURL(host, port, useTLS, code)
        }

        private fun parseQueryParams(query: String): Map<String, String>? {
            return try {
                query.split("&").mapNotNull {
                    val i = it.indexOf('=')
                    if (i < 0) null else {
                        val key = URLDecoder.decode(it.substring(0, i), "UTF-8")
                        val value = URLDecoder.decode(it.substring(i + 1), "UTF-8")
                        key to value
                    }
                }.toMap()
            } catch (e: IllegalArgumentException) {
                // URLDecoder.decode throws IllegalArgumentException on malformed
                // percent-encoding. While URI() typically rejects these first,
                // we guard here for defense-in-depth.
                null
            }
        }
    }
}
